exec/bigmachine.go

```diff
@@ -551,38 +567,40 @@ func (w *worker) FuncLocations(ctx context.Context, _ struct{}, locs *[]string)
 	return nil
 }
 
-type compileRequest struct {
-	Inv bigslice.Invocation
-	Env CompileEnv
-}
-
 // Compile compiles an invocation on the worker and stores the
 // resulting tasks. Compile is idempotent: it will compile each
 // invocation at most once.
-func (w *worker) Compile(ctx context.Context, req compileRequest, _ *struct{}) (err error) {
+func (w *worker) Compile(ctx context.Context, invReader io.Reader, _ *struct{}) (err error) {
 	defer func() {
 		if e := recover(); e != nil {
 			err = fmt.Errorf("invocation panic! %v", e)
 			err = errors.E(errors.Fatal, err)
 		}
 	}()
-	return w.compiles.Do(req.Inv.Index, func() error {
+	dec := gob.NewDecoder(invReader)
+	var inv execInvocation
+	if err = dec.Decode(&inv); err != nil {
+		// We may want to distinguish read errors from decode errors and
```

Comment:
Optional: At first I thought this would be important to disambiguate, but then I decided since we control encoding (and check errors), it's very unlikely that there's invalid data here. If you agree, consider stating that in this comment, and remove the TODO-like language.

Reply:
We control encoding, but I don't think it's type-checked anywhere. I think the way we'd get invalid data is by an accident like:
```go
m.RetryCall(ctx, "Worker.Compile", readerThatDoesntHoldAnInvocation, nil)
```
Thinking about this a bit more, I think the existing code conflates these cases:

- [Server returns 400](https://github.com/grailbio/bigmachine/blob/a63391e525386b89ccc88a57c0b90e31726b1b54/rpc/server.go#L234) .
- [Client passes back fatal error](https://github.com/grailbio/bigmachine/blob/a63391e525386b89ccc88a57c0b90e31726b1b54/rpc/client.go#L242-L244)

I'm now thinking that this may never matter because virtually all cases where these reads could fail are handled on the client side as a failure of the HTTP call itself.

Reply:
Okay. I just deleted the comment, as I think it doesn't really add anything and likely won't be an issue anyway.